app/controllers/tests_controller.rb

```diff
@@ -2,30 +2,39 @@ class TestsController < ApplicationController
   skip_before_action :check_admin, except: [:index]
 
   def create
-    course = Problem.find(params[:problem_id]).course
+    problem = Problem.includes(assignment: :course).find(params[:problem_id])
+    course = problem.course
     if !current_user_in_course?(course)
       head :forbidden
       return
     end
 
     uploaded_file = params[:file]
 
+    test = Test.new(
+      :user_id => current_user.id,
+      :name => params[:name],
+      :hint => params[:hint],
+      :problem_id => params[:problem_id],
+    )
     ActiveRecord::Base.transaction do
       file = DbFile.create(
         name: uploaded_file.original_filename,
         content_type: 'text/plain',
         contents: uploaded_file.read,
       )
 
-      test = Test.create(
-        :user_id => current_user.id,
-        :name => params[:name],
-        :hint => params[:hint],
-        :problem_id => params[:problem_id],
-        :db_file_id => file.id,
-      )
+      test.db_file_id = file.id
+      test.save!
+    end
+
+    render status: :created, json: test
+
+    FillExpectedOutputJob.perform_later(test.id)
 
-      render status: :created, json: test
+    # Rerun all previous submissions on this new test case
```

Comment:
If it isn't too much work, let's do that; but if not, we can leave it for a later iteration.